File: drivers/java/src/main/java/dev/reddb/Conn.java
package dev.reddb;

import java.util.List;

/**
 * Connection-shaped surface every transport (RedWire, HTTP) implements.
 * All methods are blocking; they return raw JSON bytes so the caller
 * can deserialise with whatever object mapper they prefer.
 */
public interface Conn extends AutoCloseable {
    /** Run a SQL query. Returns the engine's JSON envelope as bytes. */
    byte[] query(String sql);

    /** Insert a single row into a collection. `payload` is anything Jackson can serialise. */
    void insert(String collection, Object payload);

    /** Insert many rows in one round trip. Each row is anything Jackson can serialise. */
    void bulkInsert(String collection, List<?> rows);

    /** Fetch one row by id. Returns the JSON envelope (`{ ok, found, ... }`) as bytes. */
    byte[] get(String collection, String id);

    /** Delete one row by id. */
    void delete(String collection, String id);

    /** Round-trip a Ping → Pong. Throws on protocol errors. */
    void ping();

    /** Idempotent close. */
    @Override
    void close();
}

File: drivers/kotlin/src/main/kotlin/dev/reddb/Conn.kt
package dev.reddb

/**
 * Connection-shaped surface every transport (RedWire, HTTP) implements.
 * Every operation is a `suspend fun`; the implementation owns one socket
 * (or one HTTP client) and serialises concurrent calls internally.
 *
 * Methods that return result rows hand back the engine's raw JSON envelope
 * as a [ByteArray] — callers deserialise with whichever object mapper they
 * prefer (Jackson, kotlinx-serialization, etc.).
 */
public interface Conn : AutoCloseable {
    /** Run a SQL query. Returns the engine's JSON envelope as bytes. */
    public suspend fun query(sql: String): ByteArray

    /** Insert a single row into a collection. `payload` is anything Jackson can serialise. */
    public suspend fun insert(collection: String, payload: Any)

    /** Insert many rows in one round trip. */
    public suspend fun bulkInsert(collection: String, rows: List<Any?>)

    /** Fetch one row by id. Returns the JSON envelope (`{ ok, found, ... }`) as bytes. */
    public suspend fun get(collection: String, id: String): ByteArray

    /** Delete one row by id. */
    public suspend fun delete(collection: String, id: String)

    /** Round-trip a Ping → Pong (or GET /admin/health on HTTP). Throws on protocol errors. */
    public suspend fun ping()

    /** Idempotent close — second call is a no-op. */
    public override fun close()
}
